src/cascade/model/model_reader.py

```diff
@@ -0,0 +1,313 @@
+        for key, smooth_value in group.items():
+            if group_name == "random_effect":
+                child_node = key[1]
+                inverted_smooth[(smooth_value, child_node)] = (group_name, key)
+            else:
+                inverted_smooth[(smooth_value, parent_node)] = (group_name, key)
+
+    return inverted_smooth
+
+
+def _read_rate_smooths(child_node, nslist_pair_table, rate_table, smooths):
+    for rate_row in rate_table.itertuples():
+        if rate_row.parent_smooth_id is not None and not isnan(rate_row.parent_smooth_id):
+            smooths.rate[rate_row.rate_name] = int(rate_row.parent_smooth_id)
+        if rate_row.child_smooth_id is not None and not isnan(rate_row.child_smooth_id):
+            # Random effects can have children with different fields but same smoothing.
+            for child in child_node:
+                smooths.random_effect[(rate_row.rate_name, child)] = int(rate_row.child_smooth_id)
+        if rate_row.child_nslist_id is not None and not isnan(rate_row.child_nslist_id):
+            child_df = nslist_pair_table[nslist_pair_table.nslist_id == rate_row.child_nslist_id]
+            for ns_row in child_df.itertuples():
+                smooths.random_effect[(rate_row.rate_name, ns_row.node_id)] = int(ns_row.smooth_id)
+
+
+def _read_mulcov_smooths(mulcov_table, covariate_table, smooths):
+    group = dict(rate_value=smooths.alpha, meas_value=smooths.beta, meas_std=smooths.gamma)
+    for mulcov_row in mulcov_table.itertuples():
+        found_name = covariate_table.query("covariate_id == @mulcov_row.covariate_id")
+        if found_name.empty:
+            MATHLOG.error(f"Mulcov covariate id {mulcov_row.covariate_id} not found in covariate table.")
+            raise RuntimeError(f"Could not find covariate id {mulcov_row.covariate_id} in covariate table.")
+        covariate_name = str(found_name.iloc[0].covariate_name)
+        if mulcov_row.mulcov_type == "rate_value":
+            target_name = RateEnum(mulcov_row.rate_id).name
+        else:
+            target_name = IntegrandEnum(mulcov_row.integrand_id).name
+
+        MATHLOG.debug(f"Covariate={covariate_name}, target={target_name}")
+        group[mulcov_row.mulcov_type][(covariate_name, target_name)] = mulcov_row.smooth_id
+
+
+def read_parent_node(dismod_file):
+    """Get ``node_id`` for parent location."""
+    return int(dismod_file.option.query("option_name == 'parent_node_id'").option_value)
+
+
+def read_child_nodes(dismod_file, parent_node):
+    """Get node ids for child locations."""
+    return dismod_file.node[dismod_file.node.parent == parent_node].node_id.values
+
+
+def rename_node_to_location(node_table, var_ids):
+    """Given a DismodGroups based on node_id, change keys
+    to be based on location_id"""
+    node_to_location = DismodGroups()
+    for group_name, group in var_ids.items():
+        for key, smooth_value in group.items():
+            if group_name == "random_effect":
```

Comment:
Hmm. In the old model this needed to happen in a lot of other places. What about the data? That also has node/locations in it.

Reply:
There's going to be a little appy() in the data that switches location to node_id. Yup. One line.

Reply:
Ok